icu4c/source/common/mlbe.cpp

```diff
@@ -128,135 +122,169 @@ int32_t MlBreakEngine::divideUpRange(UText *inText, int32_t rangeStart, int32_t
-                          feature.setTo(u"BW2:", 4).append(elementList[2]).append(elementList[3])));
+    if (indexList[2] != -1 && indexList[3] != -1) {
+        // BW2
+        end = (indexList[4] != -1) ? indexList[4] : codeUts;
+        score += fModel[7].geti(inString.tempSubString(indexList[2], end - indexList[2]));
     }
-    if (elementList[3] != INVALID && elementList[4] != INVALID) {
-        score += (2 * fModel.geti(
-                          feature.setTo(u"BW3:", 4).append(elementList[3]).append(elementList[4])));
+    if (indexList[3] != -1 && indexList[4] != -1) {
+        // BW3
+        end = (indexList[5] != -1) ? indexList[5] : codeUts;
+        score += fModel[8].geti(inString.tempSubString(indexList[3], end - indexList[3]));
     }
-    if (elementList[0] != INVALID && elementList[1] != INVALID && elementList[2] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW1:", 4)
-                                      .append(elementList[0])
-                                      .append(elementList[1])
-                                      .append(elementList[2])));
+    if (indexList[0] != -1 && indexList[1] != -1 && indexList[2] != -1) {
+        // TW1
+        end = (indexList[3] != -1) ? indexList[3] : codeUts;
+        score += fModel[9].geti(inString.tempSubString(indexList[0], end - indexList[0]));
     }
-    if (elementList[1] != INVALID && elementList[2] != INVALID && elementList[3] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW2:", 4)
-                                      .append(elementList[1])
-                                      .append(elementList[2])
-                                      .append(elementList[3])));
+    if (indexList[1] != -1 && indexList[2] != -1 && indexList[3] != -1) {
+        // TW2
+        end = (indexList[4] != -1) ? indexList[4] : codeUts;
+        score += fModel[10].geti(inString.tempSubString(indexList[1], end - indexList[1]));
     }
-    if (elementList[2] != INVALID && elementList[3] != INVALID && elementList[4] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW3:", 4)
-                                      .append(elementList[2])
-                                      .append(elementList[3])
-                                      .append(elementList[4])));
+    if (indexList[2] != -1 && indexList[3] != -1 && indexList[4] != -1) {
+        // TW3
+        end = (indexList[5] != -1) ? indexList[5] : codeUts;
+        score += fModel[11].geti(inString.tempSubString(indexList[2], end - indexList[2]));
     }
-    if (elementList[3] != INVALID && elementList[4] != INVALID && elementList[5] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW4:", 4)
-                                      .append(elementList[3])
-                                      .append(elementList[4])
-                                      .append(elementList[5])));
+    if (indexList[3] != -1 && indexList[4] != -1 && indexList[5] != -1) {
```

Comment:
after your last PR, TW1 to TW3 (not sure about TW4, please look into that) could be merged into one for loop
for (int32_t j = 0; j < 3; j++) {
  if (indexList[j] != -1 && indexList[j+1] != -1 && indexList[j+2] != -1) {
        // TW(j+1)
        end = (indexList[j+3] != -1) ? indexList[j+3] : codeUts;
        score += fModel[j+9].geti(inString.tempSubString(indexList[j], end - indexList[j]));
    }
}

Reply:
Done